src/main/scala/com/scality/clueso/merge/TableFilesMerger.scala

```diff
@@ -0,0 +1,200 @@
+package com.scality.clueso.merge
+
+import java.io.File
+import java.net.URI
+import java.util.{Date, UUID}
+
+import com.scality.clueso.{CluesoConfig, CluesoConstants}
+import com.typesafe.config.ConfigFactory
+import org.apache.hadoop.conf.{Configuration => HadoopConfig}
+import org.apache.hadoop.fs.{FileSystem, Path, PathFilter}
+import org.apache.spark.sql.{Dataset, Row, SparkSession}
+
+class TableFilesMerger(config: CluesoConfig) {
+  val parquetFilesFilter = new PathFilter {
+    override def accept(path: Path): Boolean = path.getName.endsWith(".parquet")
+  }
+
+  def hadoopConfig(config: CluesoConfig): HadoopConfig = {
+    val c = new HadoopConfig()
+    c.set("fs.s3a.connection.ssl.enabled", config.s3SslEnabled)
+    c.set("fs.s3a.endpoint", config.s3Endpoint)
+    c.set("fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
+    c.set("fs.s3a.access.key", config.s3AccessKey)
+    c.set("fs.s3a.secret.key", config.s3SecretKey)
+    c
+  }
+
+  val fs = FileSystem.get(new URI(config.mergePath), hadoopConfig(config))
+
+  val spark = SparkSession
+    .builder
+    .config("spark.hadoop.fs.s3a.connection.ssl.enabled", config.s3SslEnabled)
+    .config("spark.hadoop.fs.s3a.endpoint", config.s3Endpoint)
+    .config("spark.hadoop.fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
+    .config("spark.hadoop.fs.s3a.access.key", config.s3AccessKey)
+    .config("spark.hadoop.fs.s3a.secret.key", config.s3SecretKey)
+    .master("local[*]")
+    .appName("Table Files Merger")
+    .getOrCreate()
+
+
+  case class MergeInstructions(numPartitions : Int)
+
+  def checkMergeEligibility(getPath: Path): Option[MergeInstructions] = {
+    // number of records criteria
+    val numberOfFiles = fs.listStatus(getPath, parquetFilesFilter).length
+
+    if (numberOfFiles >= config.mergeMinFiles) {
+      val data = spark.read
+          .parquet(getPath.toUri.toString)
+
```

Comment:
Merging operation is expensive as has to merge existing with new by reading all. It's currently doing on a per partition basis, so it doesn't have to read the full table.

The ingestion pipeline is writing a file per kafka partition, for each S3 bucket we have an event for, every 10 seconds. While it's a pain to open up these many files in case of query, we should make sure that whenever we run Merge, it's worth it. There are multiple parameters in place that we will have to tune when testing with a real scenario to see what's the best compromise between query time and number of merges the cluster can sustain.